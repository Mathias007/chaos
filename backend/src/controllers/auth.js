import jwt from "jsonwebtoken";
// import bcrypt from 'bcrypt'
import config from "../config/config";
import resStatuses from "../config/resStatuses";

import UserSchema from "../models/users";

const { TOKEN_SECRET_JWT } = config;
const { UNAUTHORIZED, NOT_FOUND } = resStatuses;

// Generate token
const generateTokens = (req, user) => {
    const ACCESS_TOKEN = jwt.sign(
        {
            sub: user._id,
            rol: user.role,
            type: "ACCESS_TOKEN"
        },
        TOKEN_SECRET_JWT,
        {
            expiresIn: 120
        }
    );
    const REFRESH_TOKEN = jwt.sign(
        {
            sub: user._id,
            rol: user.role,
            type: "REFRESH_TOKEN"
        },
        TOKEN_SECRET_JWT,
        {
            expiresIn: 480
        }
    );
    return {
        accessToken: ACCESS_TOKEN,
        refreshToken: REFRESH_TOKEN
    };
};

// Controller login user
exports.loginUser = (req, res, next) => {
    const messages = {
        CASE_UNAUTHORIZED_MESSAGE: "Wystąpił problem z autoryzacją!",
        CASE_NOT_FOUND_MESSAGE: "Nie znaleziono użytkownika!"
        // CASE_INVALID_DATA: "Dane logowania są niepoprawne!",
        // CASE_SUCCESS_MESSAGE: "Logowanie zakończyło się sukcesem."
    };

    const {
        CASE_UNAUTHORIZED_MESSAGE,
        CASE_NOT_FOUND_MESSAGE
        // CASE_INVALID_DATA,
        // CASE_SUCCESS_MESSAGE
    } = messages;

    UserSchema.findOne(
        {
            name: req.body.name,
            password: req.body.password
        },
        (err, user) => {
            console.log(user);
            if (err || !user) {
                res.status(UNAUTHORIZED).send({
                    message: CASE_UNAUTHORIZED_MESSAGE
                });
                next(err);
            } else if (!user) {
                res.status(NOT_FOUND).send({
                    message: CASE_NOT_FOUND_MESSAGE
                });
                // if (bcrypt.compareSync(req.body.password, user.password)) {
                // } else {
                // res.status(CASE_UNAUTHORIZED_MESSAGE).send({
                //     message: CASE_INVALID_DATA
                // })
                // }
            } else {
                console.log(user);
                res.json(generateTokens(req, user));
            }
        }
    ).select("password");
};

// Verify accessToken
exports.accessTokenVerify = (req, res, next) => {
    const messages = {
        CASE_TOKEN_MISSING: "Token dostępowy zaginął w akcji!",
        CASE_TOKEN_INCOMPLETE: "Token dostępowy jest niekompletny!",
        CASE_TOKEN_INVALID: "Token dostępowy jest niepoprawny!"
    };

    const {
        CASE_TOKEN_MISSING,
        CASE_TOKEN_INCOMPLETE,
        CASE_TOKEN_INVALID
    } = messages;

    if (!req.headers.authorization) {
        return res.status(UNAUTHORIZED).send({
            message: CASE_TOKEN_MISSING
        });
    }
    const BEARER = "Bearer";
    const AUTHORIZATION_TOKEN = req.headers.authorization.split(" ");
    if (AUTHORIZATION_TOKEN[0] !== BEARER) {
        return res.status(UNAUTHORIZED).send({
            message: CASE_TOKEN_INCOMPLETE
        });
    }
    jwt.verify(AUTHORIZATION_TOKEN[1], TOKEN_SECRET_JWT, function(err) {
        if (err) {
            return res.status(UNAUTHORIZED).send({
                message: CASE_TOKEN_INVALID
            });
        }
        next();
    });
};

// Verify refreshToken
exports.refreshTokenVerify = (req, res, next) => {
    const messages = {
        CASE_TOKEN_MISSING: "Token odświeżeniowy zaginął w akcji!",
        CASE_TOKEN_INCOMPLETE: "Token odświeżeniowy jest niekompletny!",
        CASE_TOKEN_INVALID: "Token odświeżeniowy jest niepoprawny!",
        CASE_USER_NOT_FOUND: "Użytkownik nie istnieje!"
    };

    const {
        CASE_TOKEN_MISSING,
        CASE_TOKEN_INCOMPLETE,
        CASE_TOKEN_INVALID,
        CASE_USER_NOT_FOUND
    } = messages;
    if (!req.body.refreshToken) {
        res.status(UNAUTHORIZED).send({
            message: CASE_TOKEN_MISSING
        });
    }
    const BEARER = "Bearer";
    const REFRESH_TOKEN = req.body.refreshToken.split(" ");
    if (REFRESH_TOKEN[0] !== BEARER) {
        return res.status(UNAUTHORIZED).send({
            message: CASE_TOKEN_INCOMPLETE
        });
    }
    jwt.verify(REFRESH_TOKEN[1], TOKEN_SECRET_JWT, function(err, payload) {
        if (err) {
            return res.status(UNAUTHORIZED).send({
                message: CASE_TOKEN_INVALID
            });
        }
        UserSchema.findById(payload.sub, function(err, person) {
            if (!person) {
                return res.status(UNAUTHORIZED).send({
                    message: CASE_USER_NOT_FOUND
                });
            }
            return res.json(generateTokens(req, person));
        });
    });
};
