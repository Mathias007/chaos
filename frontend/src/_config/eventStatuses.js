const eventStatuses = {
    articles: {
        ARTICLE_SUCCESSFULLY_LOADED: "ARTICLE_SUCCESSFULLY_LOADED",
        ARTICLE_NOT_FOUND: "ARTICLE_NOT_FOUND",
        ARTS_LIST_LOADED: "ARTS_LIST_LOADED",
        ARTS_LIST_NOT_FOUND: "ARTS_LIST_NOT_FOUND",
        ARTICLE_ADDED: "ARTICLE_ADDED",
        ARTICLE_ADDING_FAILED: "ARTICLE_ADDING_FAILED",
        ARTICLE_SUCCESFULLY_EDITED: "ARTICLE_SUCCESFULLY_EDITED",
        ARTICLE_EDITING_FAILED: "ARTICLE_EDITING_FAILED",
        ARTICLE_SUCCESFULLY_DELETED: "ARTICLE_SUCCESFULLY_DELETED",
        ARTICLE_DELETING_FAILED: "ARTICLE_DELETING_FAILED",
        AUTHENTICATION_ERROR: "AUTHENTICATION_ERROR",
        RESET_STATUS: "RESET_STATUS"
    },
    auth: {
        PASSWORD_TOO_SHORT: "PASSWORD_TOO_SHORT",
        PASSWORD_INCORRECT: "PASSWORD_INCORRECT",
        PASSWORDS_NOT_MATCH: "PASSWORDS_NOT_MATCH",
        USER_LOADED: "USER_LOADED",
        LOGIN_SUCCESSFUL: "LOGIN_SUCCESSFUL",
        AUTHENTICATION_ERROR: "AUTHENTICATION_ERROR",
        LOGIN_FAILED: "LOGIN_FAILED",
        LOGOUT_SUCCESSFUL: "LOGOUT_SUCCESSFUL",
        RESET_STATUS: "RESET_STATUS"
    },
    news: {
        NEWS_SUCCESSFULLY_LOADED: "NEWS_SUCCESSFULLY_LOADED",
        NEWS_NOT_FOUND: "NEWS_NOT_FOUND",
        NEWS_LIST_LOADED: "NEWS_LIST_LOADED",
        NEWS_LIST_NOT_FOUND: "NEWS_LIST_NOT_FOUND",
        NEWS_ADDED: "NEWS_ADDED",
        NEWS_ADDING_FAILED: "NEWS_ADDING_FAILED",
        NEWS_SUCCESFULLY_EDITED: "NEWS_SUCCESFULLY_EDITED",
        NEWS_EDITING_FAILED: "NEWS_EDITING_FAILED",
        NEWS_SUCCESFULLY_DELETED: "NEWS_SUCCESFULLY_DELETED",
        NEWS_DELETING_FAILED: "NEWS_DELETING_FAILED",
        AUTHENTICATION_ERROR: "AUTHENTICATION_ERROR",
        RESET_STATUS: "RESET_STATUS"
    },
    knights: {
        KAM_BUILDINGS_LOADED: "KAM_BUILDINGS_LOADED",
        KAM_BUILDINGS_NOT_FOUND: "KAM_BUILDINGS_NOT_FOUND",
        KAM_UNITS_LOADED: "KAM_UNITS_LOADED",
        KAM_UNITS_NOT_FOUND: "KAM_UNITS_NOT_FOUND",
        AUTHENTICATION_ERROR: "AUTHENTICATION_ERROR"
    },
    tzar: {
        TZAR_NATIONS_LOADED: "TZAR_NATIONS_LOADED",
        TZAR_NATIONS_NOT_FOUND: "TZAR_NATIONS_NOT_FOUND",
        TZAR_UNITS_LOADED: "TZAR_UNITS_LOADED",
        TZAR_UNITS_NOT_FOUND: "TZAR_UNITS_NOT_FOUND",
        TZAR_SPELLS_LOADED: "TZAR_SPELLS_LOADED",
        TZAR_SPELLS_NOT_FOUND: "TZAR_SPELLS_NOT_FOUND",
        TZAR_TECHNOLOGIES_LOADED: "TZAR_TECHNOLOGIES_LOADED",
        TZAR_TECHNOLOGIES_NOT_FOUND: "TZAR_TECHNOLOGIES_NOT_FOUND",
        AUTHENTICATION_ERROR: "AUTHENTICATION_ERROR"
    },
    users: {
        USER_SUCCESSFULLY_LOADED: "USER_SUCCESSFULLY_LOADED",
        USER_NOT_FOUND: "USER_NOT_FOUND",
        USERS_LIST_LOADED: "USERS_LIST_LOADED",
        USERS_LIST_NOT_FOUND: "USERS_LIST_NOT_FOUND",
        USER_ADDED: "USER_ADDED",
        USER_ADDING_FAILED: "USER_ADDING_FAILED",
        USER_SUCCESFULLY_EDITED: "USER_SUCCESFULLY_EDITED",
        USER_EDITING_FAILED: "USER_EDITING_FAILED",
        USER_SUCCESFULLY_DELETED: "USER_SUCCESFULLY_DELETED",
        USER_DELETING_FAILED: "USER_DELETING_FAILED",
        AUTHENTICATION_ERROR: "AUTHENTICATION_ERROR",
        RESET_STATUS: "RESET_STATUS"
    }
};

export default eventStatuses;
