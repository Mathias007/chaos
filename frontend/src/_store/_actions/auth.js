import eventStatuses from "../../_config/eventStatuses";

import fetchLogin from "./services/LoginService";

const {
    USER_LOADED,
    LOGIN_SUCCESSFUL,
    LOGIN_FAILED,
    LOGOUT_SUCCESSFUL,
    AUTHENTICATION_ERROR,
    RESET_STATUS
} = eventStatuses.auth;

export const login = (name, password, stayLogged) => {
    return (dispatch, getState) => {
        const dispatchLoginSuccessful = function(response, name, stayLogged) {
            dispatch({
                type: LOGIN_SUCCESSFUL,
                data: response.data,
                name,
                stayLogged,
                status: response.status
            });
            dispatch({
                type: USER_LOADED,
                name: name
            });

            return response.data, name, stayLogged;
        };

        const dispatchUserAuthError = function(response) {
            dispatch({
                type: AUTHENTICATION_ERROR,
                data: response.data,
                status: response.status
            });

            throw response.data;
        };

        const dispatchLoginFailed = function(response) {
            dispatch({
                type: LOGIN_FAILED,
                data: response.data,
                status: response.status
            });
            throw response.data;
        };

        return fetchLogin(
            name,
            password,
            stayLogged,
            dispatchLoginSuccessful,
            dispatchUserAuthError,
            dispatchLoginFailed
        );
    };
};

export const logout = () => {
    return (dispatch, getState) => {
        return dispatch({
            type: LOGOUT_SUCCESSFUL
        });
    };
};

export const cleanServerStatus = () => {
    return (dispatch, getState) => {
        return dispatch({
            type: RESET_STATUS
        });
    };
};
