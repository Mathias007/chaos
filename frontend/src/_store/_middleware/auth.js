import eventStatuses from "../../_config/eventStatuses";

const { USER_LOADED, LOGIN_SUCCESSFUL, LOGOUT_SUCCESSFUL } = eventStatuses.auth;

const persistTokens = store => next => action => {
    switch (action.type) {
        case USER_LOADED:
            console.log(store.getState().auth);
            break;

        case LOGIN_SUCCESSFUL:
            localStorage.setItem("accessToken", action.data.accessToken);
            localStorage.setItem("refreshToken", action.data.refreshToken);
            localStorage.setItem("remember", action.remember);
            localStorage.setItem("name", action.name);
            break;

        case LOGOUT_SUCCESSFUL:
            localStorage.removeItem("accessToken");
            localStorage.removeItem("refreshToken");
            localStorage.removeItem("remember");
            localStorage.removeItem("name");
            break;

        default:
            break;
    }

    return next(action);
};

export default persistTokens;
