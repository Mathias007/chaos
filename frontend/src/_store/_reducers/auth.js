import eventStatuses from "../../_config/eventStatuses";

import * as Auth from '../authFunctions';

const {
    REGISTRATION_SUCCESFULL,
    REGISTRATION_ERROR,
    REGISTRATION_FAILED,
    USER_LOADED,
    LOGIN_SUCCESSFUL,
    AUTHENTICATION_ERROR,
    LOGIN_FAILED,
    LOGOUT_SUCCESSFUL,
    SET_AUTHENTICATED,
    SET_AUTH_USER,
    SET_ID_REPRESENTING_TOKEN_REFRESH_COUNTER,
    SET_USERS_LIST
} = eventStatuses.auth;

const initialState = {
    // old auth
    accessToken: localStorage.getItem("accessToken"),
    refreshToken: localStorage.getItem("refreshToken"),
    isAuthenticated: null,
    name: localStorage.getItem("name"),
    errorMessage: "",

    // new auth (with refresh)
    isAuthorized: false,
    tokenRefreshCounterId: null,
    auth: {},
    account: {

    },
    users: []
};

export default function auth(state = initialState, action) {
    switch (action.type) {
        // new auth
        case SET_AUTH_USER:
            // (tokens) 
            const decodeAccessToken = Auth.decodeJWT(action.data && action.data.accessToken);
            return {
                ...state,
                ...action.data,
                auth: {
                    accessToken: action.data && action.data.accessToken || "",
                    refreshToken: action.data && action.data.refreshToken || "",
                    sub: decodeAccessToken && decodeAccessToken.sub || "",
                    rol: decodeAccessToken && decodeAccessToken.rol || "",
                    iat: decodeAccessToken && decodeAccessToken.iat || "",
                    exp: decodeAccessToken && decodeAccessToken.exp || ""
                }
            }

        case SET_AUTHENTICATED:
            console.log(state);
            // (isAuth) 
            return {
                ...state,
                ...action.data,
                isAuthorized: action.data
            }

        case SET_ID_REPRESENTING_TOKEN_REFRESH_COUNTER:
            // (state, payload) {
            return {
                ...state,
                ...action.data,
                tokenRefreshCounterId: action.data
            }

        case SET_USERS_LIST:
            // (state, payload)
            return {
                ...state,
                ...action.data,
                users: action.data.users
            }

        // old auth
        case USER_LOADED:
            return {
                ...state,
                isAuthenticated: true
            };

        case REGISTRATION_SUCCESFULL:
            console.log(action.data);
            return {
                ...state,
                ...action.data,
                autoLogin: action.stayLogged,
                name: localStorage.getItem("name")
            };

        case REGISTRATION_ERROR:
            return {
                ...state,
                ...action.data,
                errorMessage: "Something went wrong."
            };

        case REGISTRATION_FAILED:
            return {
                ...state,
                ...action.data,
                errorMessage: "Registration data is incorrect."
            };

        case LOGIN_SUCCESSFUL:
            console.log(action.remember);

            return {
                ...state,
                ...action.data,
                name: localStorage.getItem("name"),
                remember: action.remember
            };

        case AUTHENTICATION_ERROR:
            return {
                ...state,
                ...action.data,
                errorMessage: "Something went wrong."
            };

        case LOGIN_FAILED:
            return {
                ...state,
                ...action.data,
                errorMessage: "Failed login or password"
            };
        case LOGOUT_SUCCESSFUL:
            return {
                ...state,
                accessToken: null,
                remember: false,
                name: null,
                isAuthenticated: false
            };

        default:
            return state;
    }
}
