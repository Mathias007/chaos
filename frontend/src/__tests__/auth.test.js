import eventStatuses from "../_config/eventStatuses";
import auth from "../_store/_reducers/auth";

const {
    USER_LOADED,
    LOGIN_SUCCESSFUL,
    LOGIN_FAILED,
    LOGOUT_SUCCESSFUL,
    AUTHENTICATION_ERROR
} = eventStatuses.auth;

const { USER_ADDED } = eventStatuses.users;

const initialState = {
    accessToken: null,
    refreshToken: null,
    autoLogin: null,
    isAuthenticated: null,
    userId: "",
    name: "",
    errorMessage: ""
};

describe("test auth reducer", () => {
    const authTokens = {
        accessToken: "aaaa.bbbb.cccc",
        refreshToken: "cccc.bbbb.aaaa",
        id: "1111-2222-3333-4444"
    };

    const name = "testUser";
    const stayLogged = true;

    const message = "Ważna wiadomość z serwera!";

    it("should return the initial state", () => {
        expect(auth(undefined, {})).toEqual(initialState);
    });

    it("should confirm user's authentication", () => {
        const action = { type: USER_LOADED };
        expect(auth(initialState, action)).toMatchSnapshot();
    });

    it("should confirm user's adding", () => {
        const action = { type: USER_ADDED, name, stayLogged };
        expect(auth(initialState, action)).toMatchSnapshot();
    });

    it("should login user", () => {
        const action = {
            type: LOGIN_SUCCESSFUL,
            data: authTokens.id,
            name,
            stayLogged
        };
        expect(auth(initialState, action)).toMatchSnapshot();
    });

    it("should get message about user's login fail", () => {
        const action = { type: LOGIN_FAILED, data: message };
        expect(auth(initialState, action)).toMatchSnapshot();
    });

    it("should logout user", () => {
        const action = { type: LOGOUT_SUCCESSFUL };
        expect(auth(initialState, action)).toMatchSnapshot();
    });

    it("should get message about a problem with authentication", () => {
        const action = { type: AUTHENTICATION_ERROR, data: message };
        expect(auth(initialState, action)).toMatchSnapshot();
    });
});
