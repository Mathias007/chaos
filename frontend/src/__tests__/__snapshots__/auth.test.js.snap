// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test auth reducer should confirm user's adding 1`] = `
Object {
  "accessToken": null,
  "autoLogin": true,
  "errorMessage": "",
  "isAuthenticated": null,
  "name": "testUser",
  "refreshToken": null,
  "userId": "",
}
`;

exports[`test auth reducer should confirm user's authentication 1`] = `
Object {
  "accessToken": null,
  "autoLogin": null,
  "errorMessage": "",
  "isAuthenticated": true,
  "name": "",
  "refreshToken": null,
  "userId": "",
}
`;

exports[`test auth reducer should get message about a problem with authentication 1`] = `
Object {
  "accessToken": null,
  "autoLogin": null,
  "errorMessage": "Ważna wiadomość z serwera!",
  "isAuthenticated": null,
  "message": "Ważna wiadomość z serwera!",
  "name": "",
  "refreshToken": null,
  "userId": "",
}
`;

exports[`test auth reducer should get message about user's login fail 1`] = `
Object {
  "accessToken": null,
  "autoLogin": null,
  "errorMessage": "Ważna wiadomość z serwera!",
  "isAuthenticated": null,
  "message": "Ważna wiadomość z serwera!",
  "name": "",
  "refreshToken": null,
  "userId": "",
}
`;

exports[`test auth reducer should login user 1`] = `
Object {
  "accessToken": "aaaa.bbbb.cccc",
  "autoLogin": true,
  "errorMessage": "",
  "id": "1111-2222-3333-4444",
  "isAuthenticated": null,
  "name": "testUser",
  "refreshToken": "cccc.bbbb.aaaa",
  "userId": undefined,
}
`;

exports[`test auth reducer should logout user 1`] = `
Object {
  "accessToken": null,
  "autoLogin": null,
  "errorMessage": "",
  "isAuthenticated": false,
  "name": null,
  "refreshToken": null,
  "remember": false,
  "userId": "",
}
`;
